sbat.c

```diff
@@ -120,4 +121,191 @@ EFI_STATUS parse_sbat(char *sbat_base, size_t sbat_size, char *buffer,
 	return efi_status;
 }
 
+EFI_STATUS
+verify_single_entry(struct sbat_entry *entry, struct sbat_var *sbat_var_entry)
+{
+	UINT16 sbat_gen, sbat_var_gen;
+
+	if (strcmp(entry->component_name, sbat_var_entry->component_name) == 0) {
+		dprint(L"component %a has a matching SBAT variable entry, verifying\n",
+			entry->component_name);
+
+		/*
+		 * atoi returns zero for failed conversion, so essentially
+		 * badly parsed component_generation will be treated as zero
+		 */
+		sbat_gen = atoi(entry->component_generation);
+		sbat_var_gen = atoi(sbat_var_entry->component_generation);
+
+		if ( sbat_gen < sbat_var_gen ) {
+			dprint(L"component %a, generation %d, was revoked by SBAT variable",
+			       entry->component_name, sbat_gen);
+			LogError(L"image did not pass SBAT verification\n");
+			return EFI_SECURITY_VIOLATION;
+		}
+	}
+	return EFI_SUCCESS;
+}
+
+static void
+clean_up_vars(list_t *entries)
+{
+	list_t *pos = NULL, *tmp = NULL;
+	struct sbat_var *entry;
+
+	for_each_sbat_var_safe(pos, tmp, entries) {
+		entry = list_entry(pos, struct sbat_var, list);
+		list_del(&entry->list);
+
+		if (entry->component_generation)
+			FreePool((CHAR8 *)entry->component_generation);
+		if (entry->component_name_size)
+			FreePool((CHAR8 *)entry->component_name_size);
+		if (entry->component_name)
+			FreePool((CHAR8 *)entry->component_name);
+		FreePool(entry);
+	}
+}
+
+EFI_STATUS
+verify_sbat(struct sbat *sbat, list_t *sbat_entries)
+{
+	unsigned int i;
+	list_t *pos = NULL;
+	EFI_STATUS efi_status = EFI_SUCCESS;
+	struct sbat_entry *entry;
+	struct sbat_var *sbat_var_entry;
+
+	if (sbat_empty(sbat_entries) || !sbat_entries) {
```

Comment:
@martinezjavier could you looks at this ? Specifically at || !sbat_entries.
This should not happen, but better check ?

Reply:
it depends if you assume the caller will never pass a NULL pointer, agreed that better to check.

But the check is backwards, it has to be `if (!sbat_entries || sbat_empty(sbat_entries)`, otherwise passing a NULL to `sbat_empty()` will cause a NULL pointer dereference when doing `head->next`.